pkg/arvo/sys/zuse.hoon

```diff
@@ -6036,6 +6037,57 @@
       ?~  myn  ~
       ?:  =(~ t.myn)  (trip i.myn)
       (weld (trip i.myn) `tape`['/' $(myn t.myn)])
+    ::
+    ::  |base16: en/decode arbitrary MSB-first hex strings
+    ::
+    ++  base16
+      ~%  %base16  +  ~
+      |%
+      ++  en
+        ~/  %en
+        |=  a=(pair @ud @)  ::  $octs, but safe
+        ^-  cord
+        ::  trailing zero bytes, possible truncated input
+        ::
+        =/  [pad=@ud dat=@]
+          =/  wid  (met 3 q.a)
+          ?:  (gte p.a wid)
+            [(sub p.a wid) q.a]
+          =/  dat  (cut 3 [0 p.a] q.a)
+          =.  wid  (met 3 dat)
+          [(sub p.a wid) dat]
+        ::
+        =/  dap  (add (mod (met 2 dat) 2) (mul 2 pad))
+        %+  rap  3
+        %+  weld  (reap dap '0')
+        %-  flop
+        %+  turn  (rip 2 dat)
+        |=  a=@C  ^-  cord
+        ?:  (lth a 10)
+          (add '0' a)
+        (add 'a' (sub a 10))
+      ::
+      ++  de
+        ~/  %de
+        |=  a=cord
+        ^-  (unit (pair @ud @))  ::  $octs, but safe
+        (rush a rule)
+      ::
+      ++  rule
+        |^  (bass 3 16 (star hit))
+        ::
+        ::  shadowed for bloq-length capture
+        ::
+        ++  bass
+          |*  [boq=@u wuc=@u tyd=^rule]
+          %+  cook
+            |=  waq=(list @)
+            =/  wid  (mul (lent waq) (dec (xeb wuc)))
+            :-  (rsh 0 boq (add wid (dec (bex boq))))
+            (roll waq |=([p=@ q=@] (add (end 0 wuc p) (mul wuc q))))
```

Comment:
Note that this fold is slightly different from `+bass` in hoon.hoon -- the `(end 0 wuc ...)` addition makes this a consistent bitwise operation, and much more suitable for jetting. I think it also captures the behavior we actually want here, and the version in hoon.hoon (along with `+boss`) should be probably be updated to match.

Note also that this is equivalent to `(repn wuc waq)`. I didn't use that because a) I'm jetting the decoding arm directly and b), I want this pattern to be used more generally, but the `+repn` jet as implemented is not suitable for large inputs.

Reply:
Would there be any sense in improving the `repn` jet directly?

Reply:
Definitely, I plan to do so.

`u3r_chop()` and the atom slab allocator show up as the hotspots in parsers like this -- the former needs to be rewritten for efficiency, and the latter can be solved by either special-casing direct-atom sizes in all the jets, or designing a slab-allocator api that can do so automatically. I aim to try for the latter.

Reply:
Given the above, this probably needs an `XX  use repn when ...`.

Reply:
This code was wrong, as were my comments: my assumptions only held if the base was a power of two. I've removed the generic code here and just written what's need for base16. I'll save the more general stuff for a better time/place.